test/common/router/router_test.cc

```diff
@@ -807,6 +812,150 @@ TEST_F(RouterTest, EnvoyUpstreamServiceTime) {
   EXPECT_TRUE(verifyHostUpstreamStats(1, 0));
 }
 
+void RouterTestBase::testAppendCluster(absl::optional<std::string> cluster_header_name) {
+  auto debug_config = std::make_unique<DebugConfig>();
+  debug_config->append_cluster_ = true;
+  if (cluster_header_name.has_value()) {
+    debug_config->cluster_header_ = *cluster_header_name;
+  }
+  callbacks_.streamInfo().filterState().setData(DebugConfig::key(), std::move(debug_config),
+                                                StreamInfo::FilterState::StateType::ReadOnly);
+
+  NiceMock<Http::MockStreamEncoder> encoder;
+  Http::StreamDecoder* response_decoder = nullptr;
+  EXPECT_CALL(cm_.conn_pool_, newStream(_, _))
+      .WillOnce(Invoke([&](Http::StreamDecoder& decoder, Http::ConnectionPool::Callbacks& callbacks)
+                           -> Http::ConnectionPool::Cancellable* {
+        response_decoder = &decoder;
+        callbacks.onPoolReady(encoder, cm_.conn_pool_.host_);
+        return nullptr;
+      }));
+  expectResponseTimerCreate();
+
+  Http::TestHeaderMapImpl headers;
+  HttpTestUtility::addDefaultHeaders(headers);
+  router_.decodeHeaders(headers, true);
+
+  Http::HeaderMapPtr response_headers(new Http::TestHeaderMapImpl{{":status", "200"}});
+  EXPECT_CALL(cm_.conn_pool_.host_->outlier_detector_, putHttpResponseCode(200));
+  EXPECT_CALL(callbacks_, encodeHeaders_(_, true))
+      .WillOnce(Invoke([&cluster_header_name](Http::HeaderMap& headers, bool) {
+        const Http::HeaderEntry* cluster_header = headers.get(Http::LowerCaseString(
+            cluster_header_name.has_value() ? *cluster_header_name : "x-envoy-cluster"));
+        EXPECT_NE(nullptr, cluster_header);
+        EXPECT_EQ("fake_cluster", cluster_header->value().getStringView());
+      }));
+  response_decoder->decodeHeaders(std::move(response_headers), true);
+  EXPECT_TRUE(verifyHostUpstreamStats(1, 0));
+}
+
+TEST_F(RouterTest, AppendCluster0) { testAppendCluster(absl::nullopt); }
+
+TEST_F(RouterTest, AppendCluster1) { testAppendCluster(absl::make_optional("x-custom-cluster")); }
+
+void RouterTestBase::testAppendUpstreamHost(absl::optional<std::string> hostname_header_name,
```

Comment:
FWIW, another way of doing this is to use `AssertionResult/Failure/Success`. This commonly provides better contextual information on the test in which an assertion fails when calling another function. In this particular case, since we have a trivial flat hierarchy, it might not be worth it, but FYI.

Reply:
Punted on this.